internal/cmd/grpc.go

```diff
@@ -0,0 +1,330 @@
+				return rdb.Shutdown(ctx).Err()
+			})
+
+			status := rdb.Ping(ctx)
+			if status == nil {
+				return nil, errors.New("connecting to redis: no status")
+			}
+
+			if status.Err() != nil {
+				return nil, fmt.Errorf("connecting to redis: %w", status.Err())
+			}
+
+			cacher = redis.NewCache(cfg.Cache, goredis_cache.New(&goredis_cache.Options{
+				Redis: rdb,
+			}))
+		}
+
+		interceptors = append(interceptors, middlewaregrpc.CacheUnaryInterceptor(cacher, logger))
+
+		logger.Debug("cache enabled", zap.Stringer("backend", cacher))
+	}
+
+	grpcOpts := []grpc.ServerOption{grpc_middleware.WithUnaryServerChain(interceptors...)}
+
+	if cfg.Server.Protocol == config.HTTPS {
+		creds, err := credentials.NewServerTLSFromFile(cfg.Server.CertFile, cfg.Server.CertKey)
+		if err != nil {
+			return nil, fmt.Errorf("loading TLS credentials: %w", err)
+		}
+
+		grpcOpts = append(grpcOpts, grpc.Creds(creds))
+	}
+
+	// initialize server
+	register.Add(fliptserver.New(logger, store))
+
+	// initialize grpc server
+	server.Server = grpc.NewServer(grpcOpts...)
+
+	// register grpcServer graceful stop on shutdown
+	server.pushShutdown(func(context.Context) error {
+		logger.Info("shutting down grpc server...")
+
+		server.Server.GracefulStop()
+
+		return nil
+	})
+
+	// register each grpc service onto the grpc server
+	register.RegisterGRPC(server.Server)
+
+	grpc_prometheus.EnableHandlingTimeHistogram()
+	grpc_prometheus.Register(server.Server)
+	reflection.Register(server.Server)
+
+	return server, nil
+}
+
+// ClientConn constructs and configures a client connection to the underlying gRPC server.
+func (s *GRPCServer) ClientConn(ctx context.Context) (*grpc.ClientConn, error) {
+	opts := []grpc.DialOption{grpc.WithBlock()}
+	switch s.cfg.Server.Protocol {
+	case config.HTTPS:
+		creds, err := credentials.NewClientTLSFromFile(s.cfg.Server.CertFile, "")
+		if err != nil {
+			return nil, fmt.Errorf("loading TLS credentials: %w", err)
+		}
+
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	case config.HTTP:
+		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	}
+
+	dialCtx, dialCancel := context.WithTimeout(ctx, 5*time.Second)
+	defer dialCancel()
+
+	return grpc.DialContext(dialCtx,
+		fmt.Sprintf("%s:%d", s.cfg.Server.Host, s.cfg.Server.GRPCPort), opts...)
+}
+
+// Run begins serving gRPC requests.
+// This methods blocks until Shutdown is called.
+func (s *GRPCServer) Run() error {
+	s.logger.Debug("starting grpc server")
+
+	return s.Server.Serve(s.ln)
+}
+
+// Shutdown tearsdown the entire gRPC stack including dependencies.
+func (s *GRPCServer) Shutdown(ctx context.Context) error {
+	s.logger.Info("shutting down GRPC server...")
+
+	// call in reverse order to emulate pop semantics of a stack
+	for i := len(s.shutdownFuncs) - 1; i >= 0; i-- {
+		if err := s.shutdownFuncs[i](ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *GRPCServer) pushShutdown(fn func(context.Context) error) {
```

Comment:
```suggestion
func (s *GRPCServer) onShutdown(fn func(context.Context) error) {
```

Nit: I feel like the structure of how the internal shutdown semantics work (stack) shouldn't matter to the caller, so wyt about just making it more clear that these functions will run on shutdown?